go-controller/pkg/kubevirt/router.go

```diff
@@ -0,0 +1,274 @@
+		// Point to point routing is no longer needed if vm
+		// is running at the node where it was created
+		if err := DeleteRoutingForMigratedPod(nbClient, pod); err != nil {
+			return err
+		}
+		return nil
+	} else {
+		// Since we are at remote zone we should not have local zone point to
+		// to point routing
+		if err := DeleteRoutingForMigratedPodWithZone(nbClient, pod, pointer.String(OvnLocalZone)); err != nil {
+			return err
+		}
+	}
+
+	podAnnotation, err := util.UnmarshalPodAnnotation(pod.Annotations, types.DefaultNetworkName)
+	if err != nil {
+		return fmt.Errorf("failed reading ovn annotation: %v", err)
+	}
+
+	node, err := watchFactory.GetNode(pod.Spec.NodeName)
+	if err != nil {
+		return err
+	}
+	transitSwitchPortAddrs, err := util.ParseNodeTransitSwitchPortAddrs(node)
+	if err != nil {
+		return err
+	}
+	for _, podIP := range podAnnotation.IPs {
+		ipFamily := utilnet.IPFamilyOfCIDR(podIP)
+		transitSwitchPortAddr, err := util.MatchFirstIPNetFamily(ipFamily == utilnet.IPv6, transitSwitchPortAddrs)
+		if err != nil {
+			return err
+		}
+		route := nbdb.LogicalRouterStaticRoute{
+			IPPrefix: podIP.IP.String(),
+			Nexthop:  transitSwitchPortAddr.IP.String(),
+			Policy:   &nbdb.LogicalRouterStaticRoutePolicyDstIP,
+			ExternalIDs: map[string]string{
+				OvnZoneExternalIDKey:                  OvnRemoteZone,
+				string(libovsdbops.VirtualMachineKey): pod.Labels[kubevirtv1.VirtualMachineNameLabel],
+				string(libovsdbops.NamespaceKey):      pod.Namespace,
+			},
+		}
+		if err := libovsdbops.CreateOrReplaceLogicalRouterStaticRouteWithPredicate(nbClient, types.OVNClusterRouter, &route, func(item *nbdb.LogicalRouterStaticRoute) bool {
+			matches := item.IPPrefix == route.IPPrefix && item.Nexthop == route.Nexthop && item.Policy != nil && *item.Policy == *route.Policy
+			return matches
+		}); err != nil {
+			return fmt.Errorf("failed adding static route at remote zone: %v", err)
+		}
+	}
+	return nil
+}
+
+func EnsureRoutingForVM(controllerName string, watchFactory *factory.WatchFactory, nbClient libovsdbclient.Client, switchNames SwitchNames, isPodScheduledinLocalZone bool, pod *corev1.Pod, networkName string) error {
+	isMigratedSourcePodStale, err := IsMigratedSourcePodStale(watchFactory, pod)
+	if err != nil {
+		return err
+	}
+	if util.PodWantsHostNetwork(pod) || !IsPodLiveMigratable(pod) || isMigratedSourcePodStale {
+		return nil
+	}
+
+	// When the live migration is finish this label will appear at the
+	// virt-launcher with the same value as pod.Spec.NodeName
+	targetNode := pod.Labels[kubevirtv1.NodeNameLabel]
+
+	// If the live migration has not yet finish but target is ready to receive
+	// traffic this annotation will be added with the readiness timestamp
+	targetReadyTimestamp := pod.Annotations[kubevirtv1.MigrationTargetReadyTimestamp]
+
+	// No live migration or target node was reached || qemu is already ready
+	if targetNode == pod.Spec.NodeName || targetReadyTimestamp != "" {
```

Comment:
Two questions:

- Why do we want to wait for this instead of adding the routing from the get-go?
- Why would we check for `targetNode == pod.Spec.NodeName` if `targetReadyTimestamp != ""` should happen first? If this is related to the copy-first or copy-last or however the live migration procedure might work, I suggest adding more detail in the comment.

Reply:
I will rework the comments, if there is no live migration the VM is not annotated wiith `targetReadyTimestamp` so we have to check the nodeNameLabel.